### ISSUE TITLE
Is.SameAs with a reference type constraint

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#2933]
### Problem

If you use `Is.SameAs(3)` or with any value type in general (enums, structs, primitive types), you get a guaranteed failure because a new object gets allocated on the heap when you pass the value type to the `object` parameter of `SameAs`. (Known as 'boxing.') On the flip side, `Is.Not.SameAs(valueTypedExpression)` is guaranteed to always succeed.

If in theory a .NET runtime decides to begin caching common boxed value types like `(object)true` or low integers (the JVM does this), then instead of a guaranteed failure, you end up simply testing a side effect of the runtime you're on.

### Guardrail

As a guardrail, what if we added an overload to require that the parameter be a reference type? That makes sure the argument expression is typed as a class, delegate, interface, or dynamic, all of which have an innate concept of reference equality:

```diff
namespace NUnit.Framework.Constraints
{
    public abstract class Is
    {
        public static SameAsConstraint SameAs(object expected);

+       public static SameAsConstraint SameAs<T>(T expected) where T : class;
    }
}
```

After this change, inherently broken `Is.SameAs` usages would become compile-time errors because the generic overload will always be considered a better match even if the constraint doesn't match:

`Is.SameAs(3)`

> The type 'int' must be a reference type in order to use it as parameter `T`

The workaround would be `Is.SameAs((object)3)` or `IsSameAs<object>(3)`, and both of those should make it clear that what you're doing is inherently meaningless because nothing can be the same as a newly-boxed, used-once value except as a runtime side-effect.

We'd want to keep the old overload to avoid a binary break. (That is, if my helper library references NUnit 3.10 and you use both my helper library and NUnit 3.11, it'd cause a runtime IL failure if my helper library references the non-generic `SameAs(object)` method.)

We could consider this for `Assert.AreSame`, too.

**Comment by CharliePoole** [comment]
@jnm2 All good points. All points any programmer writing an application and testing it should know about. Personally, I question how much effort it's worth spending to help someone who writes `Is.SameAs(3)`, since they lack some basic understanding of C#.

But maybe __we__ should, even if I don't want to. In the early days, we assumed that our users were as good at programming as we were or, if not, were in the process of becoming good at it. This idea brings up questions in my mind as to who our current audience really is... or more exactly, what audience do ~~we~~ you want to cater to. That really does make it a nice issue for discussion about the future direction of NUnit.

Usually, I don't like to broaden issues, but this one makes me want to do it because I know there are dozens of places in NUnit with semi-hidden things like this. Originally it was by design. The idea was that NUnit would try to do whatever you told it to do, even if it were something like applying reference equality to an integer. You would get an error and figure out you were doing something wrong and stop doing it.

That was a valid point of view and we applied it fairly consistently. There are other valid points of view, such as the one you are implying in this issue. I have no problem seeing the point of view change, but I think it may take a while to find all those places that need a corresponding change in the code. (Also, I'm glad I'm not having to do it. 😸 )

**Comment by jnm2** [comment]
@CharliePoole One place where I'd expect to see this happen to more-experienced devs is when refactoring a class to a struct. It's hard for me to make a call on what is a beginner versus an advanced pitfall. As a programmer myself who hopefully isn't too shabby, guardrails like this all just look like refactoring timesavers to me.

**Comment by CharliePoole** [comment]
@jnm2 Hope you noticed I didn't take a position on this!

**Comment by jnm2** [comment]
@CharliePoole Yes! I like what you had to say. I'm trying to soften the idea that this is a totally radical point of view, requiring changes in many places to maintain consistency.
I saw nearly an identical scenario play out at work this week, so it's very plausible in my mind that this isn't just a beginner issue.

**Comment by mikkelbu** [comment]
> After this change, inherently broken Is.SameAs usages would become compile-time errors because the generic overload will always be considered a better match even if the constraint doesn't match:

I'm not sure that this will always be the case, see " Bestest Betterness https://github.com/dotnet/roslyn/pull/24756. AFAICR from a talk this will remove the generic overload as a candidate as the constraint do not match (I'm on the phone, so my Google skills are somewhat limited :smile).
Personally, I don't think I've ever used `Is.SameAs`, so I don't feel that this concrete issue is very important, but I do think that it makes sense to make the API as user friendly as possible.

**Comment by ChrisMaddock** [comment]
I think this is a really good enhancement, and a good direction to be moving in in general. Good thinking with the struct -> class refactor example, Joseph.

I see no reason why the framework shouldn’t aim to stop someone shooting themselves in the foot as much as possible - in fact, I’d be very happy if beginner programmers were looking at NUnit, and considering testing early! 

This sounds to me like a valid standalone enhancement - I don’t see why we’d need to go through the entire framework and look for similar cases, unless there’s a consistently issue I’m missing? It seems to me that items like this are enhancements we can add as and when they are raised.

**Comment by ChrisMaddock** [comment]
(Good find Mikkel! The technical side of what you’re suggesting Joseph - I haven’t checked out, that’s not something I know about off the top of my head!)

**Comment by jnm2** [comment]
@mikkelbu You're right. In fact, it's already doing this in 2.8.2. [sharplab.io](https://sharplab.io/#v2:EYLgHgbALANALiATgVwHYB8ACAGABJgRgG4BYAKHMwGZ8CJ8AmWgdnIG9zcv8bD7MotABwAKAJSduHMt1m4AygEMAtgFMAggGcRVMaRncAvpK4metfoKVqtIgPbAAVqoDGcXKrAAHV3FUATCQMuaTluQgBOEQAiRWi9M2Ng3DNqC3wrFQ1NAB4AFQA+ETyPb18AsVwAdwALVURVXBKQRjNQsNoo6OB4/VkkpKA==)

Since IIRC they're taking the same approach with bestest betterness as they did with better betterness, it won't matter which langversion you specify—you'll get the new overload resolution rules no matter what.

Therefore, it wouldn't make sense long-term to add this overload unless we obsoleted the other overload with a message that only reference types make sense. What does everyone think about that?

**Comment by mbcrawfo** [comment]
I'd vote in favor of this change.  It bit me this morning thanks to a change from manually generated test data to usage of AutoFixture combined with `var`, which allowed the type of a variable to change from an interface to a struct.  I don't think it's a matter of catering to less experienced programmers, it's more about leveraging the compiler and type system to keep us from doing logically invalid behavior.  After all, boxing isn't something most C# developers deal with on a daily basis.  Even when one has a thorough understanding of the behavior it's easy for that knowledge to get pushed into a corner and hidden behind cobwebs.  😉

**Comment by CharliePoole** [comment]
I'm in favor of changing to give an error when a value type is used. I don't think we should keep both overloads.

On the more general issue... I didn't notice some of the replies earlier. I fear I was misunderstood. I'm not saying "If we make this change we will be forced to look at a lot of other things." My point is that there are a lot of things I believe should be looked at centering around the question of who our target user is and this would be a good place to start.

**Comment by manfred-brands** [comment]
We need to try and limit the `SameAsConstraint` constraint to `where TActual : class?`
Same for the `NullConstraint`

As said before we also need to ensure that these constraints don't match the generic overload.

**Comment by stevenaw** [comment]
A few things I've noticed during some investigations today:

- We've been a bit more conscious about binary-breaking changes as of late. Earlier discussions suggest removing the `SameAs(object? value)` method at the same time as introducing a new generic method would be a binary-breaking change.
- Boxed value types may be reference-equal if both expected and actual are both `default` (null). This could also apply to _nullable_ value types like `int?` (see example in https://github.com/nunit/nunit.analyzers/issues/971)
- Adding `where T : class?` generic constraint would exclude both scenarios currently supported in point 2 above. 

We could add a generic method _without_ the `where T : class?` constraint if we liked. It would start to slowly continue down a path of adding generic constraints to APIs. Existing calls using `object` either for boxed value types or just cast reference typed would continue to use the `object?` overload which complicates even marking it `Obsolete`.

EDIT: Our analyzers currently help provide compile-time protection by raising `NUnit2040` when non-reference types are passed in

**Comment by CharliePoole** [comment]
A side question comes to mind... what if you decided that this breaking change were a desirable change for the future? Is there some place you could put it? Would a long-running 5.0 branch be useful at this point?

**Comment by stevenaw** [comment]
@CharliePoole You have definitely just voiced something that's been on my own mind the last few days. 🙂
@nunit/framework-team @OsirisTerje @manfred-brands Is this a good time for us to start v5? Perhaps we could also start a discussion on it if we felt it needed
